platform/logger_internal.h

```diff
@@ -0,0 +1,273 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** \addtogroup platform */
+/** @{*/
+/**
+ * \defgroup platform_log Logger functions
+ * @{
+ */
+
+#ifndef LOGGER_INTERNAL_H
+#define LOGGER_INTERNAL_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include "platform/mbed_preprocessor.h"
+
+#ifndef MBED_CONF_MAX_LOG_STR_SIZE
+#define MBED_CONF_MAX_LOG_STR_SIZE      128
+#endif
+
+#ifndef MBED_CONF_EXTERNAL_BUFFER_SIZE
+#define MBED_CONF_EXTERNAL_BUFFER_SIZE  1024
+#endif
+
+#if defined (MBED_ID_BASED_TRACING)
+#define LOG_DATA_TYPE_   uint32_t
+#else
+#define LOG_DATA_TYPE_   char
+#endif
+
+#if defined(__ARMCC_VERSION)
+#define LOG_FILE_NAME_           __MODULE__
+#else
+#define LOG_FILE_NAME_           __BASE_FILE__
+#endif
+
+#define LOG_SINGLE_STR_SIZE_     MBED_CONF_MAX_LOG_STR_SIZE
+// Log-Level Strings
+#define LOG_GEN_                 "GEN "
+#define LOG_ERR_CRITICAL_        "CRT "
+#define LOG_ERR_                 "ERR "
+#define LOG_WARN_                "WARN"
+#define LOG_DEBUG_               "DBG "
+#define LOG_INFO_                "INFO"
+#define LOG_TRACE_               "TRAC"
+
+#define GET_LOG_STRING(ll)      ((ll > LOG_LEVEL_DEBUG) ? LOG_TRACE_ : \
```

Comment:
Embedding strings in the image is again doubtful on speed/space trade-off, and it seems to prevent you doing run-time selection of trace level, which is required.